File: python/src/threads/agents/dynamic.py
"""`dynamic_agent()` (spec/api.json `dynamicAgent`): a template for team members the lead defines
at start, within what this code pins. Pure, like `agent()`."""

import re
from collections.abc import Mapping, Sequence
from dataclasses import replace
from typing import Final, Required, Unpack

from pydantic import BaseModel

from threads.agents.bindings import AppTool, ToolServer
from threads.agents.config import ConfigError
from threads.agents.dynamic_agent import DynamicAgent
from threads.agents.factory import CommonOptions, Links, build_definition, output_model
from threads.loop.model import Model

_KEY: Final = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
_REFUSED: Final = frozenset({"team", "team_limits", "subagents", "handoffs"})


class DynamicAgentOptions(CommonOptions, total=False):
    models: Required[Mapping[str, Model]]
    """The models a starter may choose from, by key, in order: the first is the default."""
    tools: Sequence[AppTool[None] | ToolServer]
    """The most a member may be given: a start chooses a subset."""
    output: type[BaseModel]
    """The structured final output every member returns."""


def dynamic_agent(**options: Unpack[DynamicAgentOptions]) -> DynamicAgent[None, object]:
    """spec/api.json `dynamicAgent`. Pure: no I/O. Raises ConfigError invalid_config for missing,
    empty or badly keyed models, and for team, subagents or handoffs."""
    given = set[str](options)
    if given & _REFUSED:
        why = "a dynamic agent can't start or hand off to other agents"
        raise ConfigError("invalid_config", why)
    if "model" in given:
        raise ConfigError("invalid_config", "a dynamic agent takes models, not model")
    models = tuple(options.get("models", {}).items())
    if not models:
        raise ConfigError("invalid_config", "a dynamic agent needs at least one of models")
    bad = next((k for k, _ in models if not _KEY.match(k)), None)
    if bad is not None:
        why = f"models key {bad!r}: lowercase letters, digits and underscores, from a letter"
        raise ConfigError("invalid_config", why)
    tools = options.get("tools", ())
    servers = tuple(t for t in tools if isinstance(t, ToolServer))
    own = tuple(t for t in tools if not isinstance(t, ToolServer))
    output = output_model(options.get("output"))
    template = build_definition(options, models[0][1], (own, servers), output, Links())
    return DynamicAgent(replace(template, models=models))

File: typescript/packages/core/src/agent/team/dynamic.ts
import type { z } from "zod";
import { Name } from "../../log";
import type { Model } from "../../model";
import { type AgentOptions, type Models, registerAs, resolve } from "../agent";
import { ConfigError } from "../errors";
import type { DynamicAgent } from "./types";

// dynamicAgent() (spec/api.json, lane 26): a template the code writes. A start chooses only a
// label, written instructions, a subset of its tools and one of its models; everything else is
// this config, pinned for every member started from it.

/** Every agent() option but model, team, subagents, handoffs and teamLimits, plus models. */
export type DynamicAgentOptions<Deps, Output> = Omit<
  AgentOptions<Deps, Output>,
  "model" | "team" | "subagents" | "handoffs" | "teamLimits"
> & {
  /** The models a start may choose, by key; the first key is the default. */
  readonly models: Readonly<Record<string, Model>>;
};

const NESTING = ["team", "subagents", "handoffs"] as const;

export function dynamicAgent<Deps = undefined>(
  options: DynamicAgentOptions<Deps, string> & { readonly output?: undefined },
): DynamicAgent<Deps, string>;
export function dynamicAgent<Deps, Output>(
  options: DynamicAgentOptions<Deps, Output> & {
    readonly output: z.ZodType<Output>;
  },
): DynamicAgent<Deps, Output>;
/** Throws ConfigError invalid_config for missing or badly keyed models, or a team of its own. */
export function dynamicAgent<Deps, Output>(
  options: DynamicAgentOptions<Deps, Output>,
): DynamicAgent<Deps, Output> {
  const models = modelsOf(options);
  const first = models[0];
  if (first === undefined) throw new Error("modelsOf returns at least one");
  const { output: schema, ...rest } = options;
  const named = { name: options.name ?? "agent" };
  if (schema === undefined) {
    const { def } = resolve<Deps, string>(
      { ...rest, model: first[1] },
      (text) => text,
      models,
    );
    registerAs(named, def, options.approvers);
    return named;
  }
  const { def } = resolve<Deps, Output>(
    { ...options, model: first[1] },
    (_text, accepted) => schema.parse(accepted),
    models,
  );
  registerAs(named, def, options.approvers);
  return named;
}

/** The models in order, checked: non-empty, each key a name. */
function modelsOf<Deps, Output>(
  options: DynamicAgentOptions<Deps, Output>,
): Models {
  const nested = NESTING.find((k) => k in options);
  if (nested !== undefined)
    throw new ConfigError(
      "invalid_config",
      `${nested}: a dynamic agent can't start or hand off to other agents`,
    );
  if ("model" in options)
    throw new ConfigError(
      "invalid_config",
      "a dynamic agent takes models, not model: models: {key: model}, the first the default",
    );
  const models = Object.entries(options.models ?? {});
  if (models.length === 0)
    throw new ConfigError(
      "invalid_config",
      "models: a dynamic agent needs at least one model, by key; the first is the default",
    );
  const bad = models.find(([k]) => !Name.safeParse(k).success);
  if (bad !== undefined)
    throw new ConfigError(
      "invalid_config",
      `models: key ${JSON.stringify(bad[0])} must be lowercase letters, digits and underscores, starting with a letter`,
    );
  return models;
}
